(2) Quinn Introduction Chapter, Certificate Configuration Page (#920)
quinn-rs/quinn: docs/book/src/quinn/certificate.md

# Certificates

In this chapter, we discuss the configuration of the certificates that is **required** for a working Quinn connection. 

QUIC uses TLS 1.3 for authentication of connections, the server will have to provide the client with a certificate confirming its identity, 

> and the customer must be configured to trust the certificates he receives from our server. 

Reviewer: The difference between "client" and "customer" seems confusing here. The connection between "QUIC uses TLS 1.3" and "the server will have" also seems to be left out.

Author: Ah I think its a translator issue here.

## Insecure Connection

For our example use case, the easiest way to allow the client to trust our server is to disable certificate verification (don't do this in production!). 
When the `dangerous_configuration` feature flag of [rustls][3] is enabled, a client can be configured to trust any server.

Start with adding a [rustls][3] dependency with the `dangerous_configuration` feature flag to your `Cargo.toml` file.

```toml
quinn = "*"
rustls = { version = "*", features = ["dangerous_configuration", "quic"] }
``` 

> Then, you can skip the certificate validation on the client by implementing [ServerCertVerifier][ServerCertVerifier] and let it assert true for any server. 

Reviewer: ```suggestion
Then, you can skip the certificate validation on the client by implementing [ServerCertVerifier][ServerCertVerifier] and let it assert verification for any server. 
```

```rust
// Implementation of `ServerCertVerifier` that verifies everything as trustworthy.
struct SkipCertificationVerification;

impl rustls::ServerCertVerifier for SkipCertificationVerification {
    fn verify_server_cert(
        &self, _: &rustls::RootCertStore, _: &[rustls::Certificate], _: webpki::DNSNameRef, _: &[u8],
    ) -> Result<rustls::ServerCertVerified, rustls::TLSError> {
        Ok(ServerCertVerified::assertion())
    }
}
```

> After that, we can configure our [ClientConfig][ClientConfig] to use this new [ServerCertVerifier][ServerCertVerifier]. 

Reviewer: ```suggestion
After that, we can configure our [ClientConfig][ClientConfig] to use this new [ServerCertVerifier][ServerCertVerifier] implementation.
```

```rust
pub fn insecure() -> ClientConfig {
    let mut cfg = quinn::ClientConfigBuilder::default().build();

    // Get a mutable reference to the 'crypto' config in the 'client config'..
    let tls_cfg: &mut rustls::ClientConfig =
        std::sync::Arc::get_mut(&mut cfg.crypto).unwrap();

    // Change the certification verifier.
    // This is only available when compiled with 'dangerous_configuration' feature.
    tls_cfg
        .dangerous()
        .set_certificate_verifier(Arc::new(SkipCertificationVerification));
    cfg
}
```
 
Finally, if you plug this [ClientConfig][ClientConfig] into the [EndpointBuilder::default_client_config()][default_client_config] your client endpoint should verify all connections as trustworthy.

## Using Certificates

In this section we look at certifying an endpoint with a real certificate. 

> This can be done with either a real certificate or a self-identified certificate.

Reviewer: Here you talk about "a real certificate", which you then split into "a real certificate or a self-identified certificate". Should probably be more explicit about what the non-self-identified real certificate is.

### Self Signed

> A [self-signed][5] certificate entails that the user signing it will be its own CA. 
> These certificates are easy to create and cost no money. 

Reviewer: ```suggestion
Relying on [self-signed][5] certificates means that clients allow servers to sign their own certificates. This is simpler because no third party has to be involved in signing the server's certificate.
```

> However, they do not offer all the security features that certificates from trusted CA's do have. 

Reviewer: ```suggestion
However, they do not offer all the security features that certificates from trusted CAs do have. 
```

> Some ways to create a self-signed certificate is by using [rcgen][4] or openssl. 

Reviewer: ```suggestion
Self-signed certificates, among other options, can be created using the [rcgen][4] crate or the openssl binary.
```

    fs::write(&cert_path, &serialized_certificate).context("failed to write certificate")?;
    fs::write(&key_path, &serialized_key).context("failed to write private key")?;

    let cert = quinn::Certificate::from_der(&cert)?;
    let key = quinn::PrivateKey::from_der(&serialized_key)?;
    Ok((cert, key))
}
```

*Note that [generate_simple_self_signed][generate_simple_self_signed] returns a [Certificate][2] that can be serialized to both `.der` and `.pem` formats.*

### Official Certificates

[Let's Encrypt][6] is a well-known Certificate Authority ([CA][1]) (certificate issuer) and distributes certificates for free.
For this section lets-encrypt is used however any CA could be used interchangeably. 

**Generate Certificate**

Let's encrypt uses [Certbot][7] to generate certificates. 
The certbot websites give clear instructions on how to use the tool.  
Normally a certificate is generated to secure a web server, however, because we generate a certificate for a protocol,
the configuration process will be slightly different than normal.
If you do want to use an existing web server to generate certificates, please follow the instructions on certbot's website.

> Select on the certbot website that you do not have a web server, we assume you dont, and follow the given installation instructions.

Reviewer: ```suggestion
Select on the certbot website that you do not have a web server (we assume you don't) and follow the given installation instructions.
```

> Let's encrypt uses [Certbot][7] to generate certificates. 
> The certbot websites give clear instructions on how to use the tool.  
> Normally a certificate is generated to secure a web server, however, because we generate a certificate for a protocol,
> the configuration process will be slightly different than normal.
> If you do want to use an existing web server to generate certificates, please follow the instructions on certbot's website.
> 
> Select on the certbot website that you do not have a web server (we assume you don't) and follow the given installation instructions.
> Certbot must answer a cryptographic challenge of the Let's Encrypt API to prove that you control the domain. 
> It uses ports 80 (HTTP) or 443 (HTTPS) to achieve this. Open the appropriate port in your firewall and router.
> 
> If certbot is installed, run `certbot certonly --standalone`, this command will start a web server in the background and start the challenge.
> Certbot asks for the required data and writes the certificate to `cert.pem` and the private key to `privkey.pem`.  
> These files can then be referenced in code.  

Reviewer: ```suggestion
[certbot][7] can be used with Let's Encrypt to generate certificates; its website comes with clear instructions.
Because we're generating a certificate for an internal test server, the process used will be slightly different compared to what you would do when generating certificates for an existing (public) website.

On the certbot website, select that you do not have a public web server and follow the given installation instructions.
certbot must answer a cryptographic challenge of the Let's Encrypt API to prove that you control the domain. 
It needs to listen on port 80 (HTTP) or 443 (HTTPS) to achieve this. Open the appropriate port in your firewall and router.

If certbot is installed, run `certbot certonly --standalone`, this command will start a web server in the background and start the challenge.
certbot asks for the required data and writes the certificate to `cert.pem` and the private key to `privkey.pem`.  
These files can then be referenced in code.  
```

 
```rust
pub fn read_cert_from_file() -> anyhow::Result<(quinn::Certificate, quinn::PrivateKey)> {
    // Read from certificate and key from directory.
    let (cert, key) = fs::read(&"./cert.pem").and_then(|x| Ok((x, fs::read(&"./privkey.pem")?)))?;

    // Parse to certificate chain whereafter taking the first certifcater in this chain.
    let cert = quinn::CertificateChain::from_pem(&cert)?.iter().next().unwrap().clone();
    let key = quinn::PrivateKey::from_pem(&key)?;

    Ok((quinn::Certificate::from(cert), key))
}
```

### Configuring Certificates

> Now you generated, or maybe you already had, the certificate, they need to be configured into the client and server. 

Reviewer: ```suggestion
Now that you have a valid certificate, the client and server need to be configured to use it.
```

After configuring plug the configuration into the `Endpoint`.

**Configure Server**

```rust
let mut builder = ServerConfigBuilder::default();
builder.certificate(CertificateChain::from_certs(vec![certificate]), key)?;
```

> This is the only thing you need to do for your sever to be secured. 

Reviewer: ```suggestion
This is the only thing you need to do for your server to be secured. 
```

**Configure Client**

```rust
let mut builder = ClientConfigBuilder::default();
builder.add_certificate_authority(certificate)?;    
```

This is the only thing you need to do for your client trust a server certificate. 

<br><hr>

> [Nextup](set-up-connection.md), lets have a look at how to setup a connection. 

Reviewer: ```suggestion
[Next](set-up-connection.md), let's have a look at how to setup a connection. 
```

In the following chapters we will go through the protocol using examples. 
The chapters are in order, first we look at configuring a certificate which is **required**, 
then at setting up a connection and finally at sending data about this connection. 

# Certificates

In this chapter, we discuss the configuration of the certificates that is **required** for a working Quinn connection. 

A [Certificate Authority (CA)][1] is an entity that issues digital [certificates][2]. 
These digital certificates certify ownership of a public key associated with, for example, a host, server, client, or document.

> Digital certificates ensure that users can be confident that the content actually comes from a reliable, secure source.

Author: Nice change. I edited some parts: 

QUIC uses TLS 1.3 for authentication of connections, the server will have to provide the client with a certificate confirming its identity, 
and the customer must be configured to trust the certificates he receives from our server.

**By default**, Quinn clients validate the cryptographic identity of the servers they connect to. 
This prevents an attacker from intercepting messages.
While it is great that quinn offers security by default it requires additional configuration.
This additional configuration will be the subject of this chapter. 

## Insecure Connection

A certificate is not practical for cases such as: peer-to-peer, trust-on-first-use,
deliberately insecure applications, or when the servers are not identified by the domain name. 
You can change certificate validation logic when the `dangerous_configuration` feature flag of [rustls][3] is enabled.

> Then the only thing that needs to be done is to configure the client to trust any server.

Author: Love this proposal

Start with adding a [rustls][3] dependency with the `dangerous_configuration` feature flag to your `Cargo.toml` file.

```toml
quinn = "*"
rustls = { version = "*", features = ["dangerous_configuration", "quic"] }
``` 

Then, you can skip the certificate validation on the client by implementing [ServerCertVerifier][ServerCertVerifier] and let it assert true for any server. 

```rust
// Implementation of `ServerCertVerifier` that verifies everything as trustworthy.
struct SkipCertificationVerification;

> impl rustls::ServerCertVerifier for SkipCertificationVerification {
>     fn verify_server_cert(
>         &self,
>         _roots: &rustls::RootCertStore,
>         _presented_certs: &[rustls::Certificate],
>         _dns_name: webpki::DNSNameRef,
>         _ocsp_response: &[u8],
>     ) -> Result<rustls::ServerCertVerified, rustls::TLSError> {
>         Ok(ServerCertVerified::assertion())
>     }

Reviewer: ```suggestion
impl rustls::ServerCertVerifier for SkipCertificationVerification {
    fn verify_server_cert(
        &self,
        _: &rustls::RootCertStore,
        _: &[rustls::Certificate],
        _: webpki::DNSNameRef,
        _: &[u8],
    ) -> Result<rustls::ServerCertVerified, rustls::TLSError> {
        Ok(ServerCertVerified::assertion())
    }
```

Maybe this way it fits in one line?

Author: Yea I can actualize that

    cfg
}
```
 
Finally, if you plug this [ClientConfig][ClientConfig] into the [EndpointBuilder::default_client_config()][default_client_config] your client endpoint should verify all connections as trustworthy.

## Using Certificates

In this section we look at certifying an endpoint with a real certificate. 
This can be done with either a real certificate or a self-identified certificate.

Let's define two useful functions that can dissect byte certificates and return quinn types.

```rust
pub fn parse_der(cert: Vec<u8>, private_key: Vec<u8>) -> anyhow::Result<(quinn::Certificate, quinn::PrivateKey)> {
    let cert = quinn::Certificate::from_der(&cert)?;
    let key = quinn::PrivateKey::from_der(&private_key)?;
    Ok((cert, key))
}

pub fn parse_pem(cert: Vec<u8>, private_key: Vec<u8>) -> anyhow::Result<(quinn::Certificate, quinn::PrivateKey)> {
    // Parse to certificate chain whereafter taking the first certifcater in this chain.
    let cert = quinn::CertificateChain::from_pem(&cert)?.iter().next().unwrap().clone();
    let key = quinn::PrivateKey::from_pem(&private_key)?;

    Ok((quinn::Certificate::from(cert), key))
}

> ```

Reviewer: I don't see much value in showing these kinds of utility functions in the documentations, since they're not at all relevant to Quinn itself.

Author: I think we can move this into the other methods.

There are two common certificate formats namely: `.pem` and `.der`.
The `.der` certificates are byte-coded, while `.pem` is text-coded.
You can translate one to the other by using tooling such as openssl or even within code self. 

> The code translation is shown above. 

Reviewer: This too doesn't seem very relevant.

Author: I was confused by it since certbot exports `.pem` files and rgcen and the certificates were using `.der`. However, I think we can remove this.

These certificates are easy to create and cost no money. 
However, they do not offer all the security features that certificates from a CA do have. 
Some ways to create a self-signed certificate is by using [rcgen][4] or openssl. 
In this example [rcgen][4] is used.   

Let's look at an example:

```rust
pub fn generate_self_signed_cert(cert_path: &str, key_path: &str) -> anyhow::Result<(quinn::Certificate, quinn::PrivateKey)> {
    // Generate dummy certificate.
    let certificate = rcgen::generate_simple_self_signed(vec!["localhost".into()]).unwrap();
    let serialized_key = certificate.serialize_private_key_der();
    let serialized_certificate = certificate.serialize_der().unwrap();

    // Write to files.
    fs::write(&cert_path, &serialized_certificate).context("failed to write certificate")?;
    fs::write(&key_path, &serialized_key).context("failed to write private key")?;

    parse_der(serialized_certificate, serialized_key)
}
```

*Note that [generate_simple_self_signed][generate_simple_self_signed] returns a [Certificate][2] that can be serialized to both `.der` and `.pem` formats.*

### Official Certificates

> [Let's Encrypt][6] is a CA and distributes certificates for free. 
> Its a very well-known CA used by many applications around the world.
> We can cover a detailed lets-encrypt tutorial but there is plenty of good documentation out there.  

Reviewer: ```suggestion
[Let's Encrypt][6] is a well-known Certificate Authority [maybe add that link from before here] (certificate issuer) and distributes certificates for free. Note that servers need to be reachable at a public DNS name to be able to get a Let's Encrypt certificate.
```

**Generate Certificate**

Let's Encrypt works with [Certbot][7], certbort generates the certificate for you.
Often a certificate is generated to secure a web server. 
Because we generate a certificate for a protocol, the configuration process will be slightly different than normal. 
We assume that you do not have a web server. 
Select on the certbot website that you do not have a web server and follow the given installation instructions.

If certbot is installed, execute `certbot certonly --standalone`, this command will fire up a web server in the background.
Certbot asks for your data, after entering it two `.pem` files are generated, namely `cert.pem` and `privkey.pem`. 
Next we can reference those files in the code.  

>  

Reviewer: How does this actually work? Seems to gloss over what kind of environment is needed for this to work.

Author: Above I mention: _Select on the certbot website that you do not have a web server and follow the given installation instructions._

Certbot website explains this in detail. I thought I would be unnecessarily to cover platform-specific notes since there is _plenty of good documentation to there_

Author: I rewrote this section a bit and pointed more to the existing certbot docs

Let's Encrypt use [Certbot][7] to generate certificates. 
The certbot websites provides clear instructions on how to use the tool.   
Normally a certificate is generated to secure a web server and the user is expected to have one up-and-running accessible from a public DNS.
However, since we generate a certificate for a protocol, the configuration process will be slightly different than normal.
If you do want to use an existing web server to generate certificates, please follow the instructions on the certbot website.

For this example it is expected that no web server is installed.
 Select on the certbot web site that you do not have a web server and follow the given installation instructions. 

If certbot is installed, run `certbot certonly --standalone`, this command will start a web server in the background.
Certbot asks for the required data and writes the certificate to `cert.pem` and private key to `privkey.pem`.  
These files can then be referred to in code.

Author: Please give it another go in the following review round.
